src/libs/actions/Workflow.ts

```diff
@@ -195,43 +207,135 @@ function removeApprovalWorkflow(policyID: string, approvalWorkflow: ApprovalWork
+    const parameters: RemoveWorkspaceApprovalParams = {policyID, authToken, employees: JSON.stringify(Object.values(updatedEmployees))};
     API.write(WRITE_COMMANDS.REMOVE_WORKSPACE_APPROVAL, parameters, {optimisticData, failureData, successData});
 }
 
 function setApprovalWorkflowMembers(members: Member[]) {
-    Onyx.merge(ONYXKEYS.APPROVAL_WORKFLOW, {members});
+    Onyx.merge(ONYXKEYS.APPROVAL_WORKFLOW, {members, errors: null});
 }
 
-function setApprovalWorkflowApprover(approver: Approver, index: number) {
-    if (!currentApprovalWorkflow) {
+/**
+ * Set the approver at the specified index in the current approval workflow
+ * @param approver - The new approver to set
+ * @param approverIndex - The index of the approver to set
+ * @param policyID - The ID of the policy
+ */
+function setApprovalWorkflowApprover(approver: Approver, approverIndex: number, policyID: string) {
+    const policy = allPolicies?.[`${ONYXKEYS.COLLECTION.POLICY}${policyID}`];
+
+    if (!currentApprovalWorkflow || !policy?.employeeList) {
         return;
     }
 
-    const updatedApprovers = [...currentApprovalWorkflow.approvers];
-    updatedApprovers[index] = approver;
-    Onyx.merge(ONYXKEYS.APPROVAL_WORKFLOW, {approvers: updatedApprovers});
+    const errors: Record<string, TranslationPaths | null> = {};
+
+    // Update the approver at the specified index and reset hints
+    const approvers: Array<Approver | undefined> = currentApprovalWorkflow.approvers.map((existingApprover) => {
+        if (!existingApprover) {
+            return;
+        }
+
+        return {...existingApprover, isInMultipleWorkflows: false};
+    });
+    approvers[approverIndex] = approver;
+
+    // Check if the approver forwards to other approvers and add them to the list
+    if (policy.employeeList[approver.email]?.forwardsTo) {
+        const personalDetailsByEmail = lodashMapKeys(personalDetails, (value, key) => value?.login ?? key);
+        const additionalApprovers = calculateApprovers({employees: policy.employeeList, firstEmail: approver.email, personalDetailsByEmail}).map((additionalApprover) => ({
+            ...additionalApprover,
+            isInMultipleWorkflows: true,
+        }));
+        approvers.splice(approverIndex, approvers.length, ...additionalApprovers);
+    }
+
+    // Check for circular references and reset errors
+    const updatedApprovers = approvers.map((existingApprover, index) => {
+        if (!existingApprover) {
+            return;
+        }
+
+        const hasCircularReference = approvers.slice(0, index).some((previousApprover) => existingApprover.email === previousApprover?.email);
+        if (hasCircularReference) {
+            errors[`approver-${index}`] = 'workflowsPage.approverCircularReference';
+        } else {
+            errors[`approver-${index}`] = null;
+        }
+
+        return {
+            ...existingApprover,
+            isCircularReference: hasCircularReference,
```

Comment:
@blazejkustra 

> @DylanDylann Try to setup a circular workflow, isCircularReference becomes true in that case.

Of course this variable needs to be used for our logic but I am saying `isCircularReference` in `calculateApprovers` function


https://github.com/Expensify/App/blob/87bdc246318f6c3fc81ad9d293024de687e457a2/src/libs/WorkflowUtils.ts#L44

The current logic works well because we have another logic to handle it while initializing `updatedApprovers`

Reply:
Without these lines:
```ts
const isCircularReference = currentApproverEmails.has(nextEmail); 

// If we've already seen this approver, break to prevent infinite loop
if (isCircularReference) {
  break;
}
```
The code would end up in an infinite loop, which we want to avoid.

Reply:
Also, if you have a circular reference configured in OldDot, and [approval workflows are built](https://github.com/software-mansion-labs/expensify-app-fork/blob/4c0078ae0a1c80edb86fe67977ffb7c16db7a511/src/pages/workspace/workflows/WorkspaceWorkflowsPage.tsx#L67) we'll have to break to avoid an infinite loop too.

Reply:
@blazejkustra  As I mentioned [here](https://github.com/Expensify/App/pull/46798#issuecomment-2283572680), this field will always be false

> @blazejkustra I see that this field always be false and It seems unnecessary.

Note that: In this function, we only use data from `employees` so the circular reference will not happen. And this logic is unnecessary because we have another logic to handle it

Reply:
> @blazejkustra As I mentioned https://github.com/Expensify/App/pull/46798#issuecomment-2283572680, this field will always be false

It's not actually true, when circular workflow is configured from OldDot this field becomes `true`:
<img width="1218" alt="Screenshot 2024-08-12 at 14 48 29" src="https://github.com/user-attachments/assets/93c952e1-f5cc-42fc-ac9d-014bc0e54337">

<img width="1218" alt="Screenshot 2024-08-12 at 14 51 50" src="https://github.com/user-attachments/assets/5a21d139-1baf-45e5-9e8c-098e3ba5ae76">

Reply:
Here's a video:

https://github.com/user-attachments/assets/142be9ec-7d30-4a52-b3dc-34674f110fc3

Reply:
@blazejkustra Ahh thanks. In this case, we have a circular reference in olddot

`blazej.kustra+3 -> blazej.kustra+4 -> blazej.kustra+4`